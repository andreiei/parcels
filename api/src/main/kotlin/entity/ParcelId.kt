package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class ParcelId(private val value: String) {
    override fun toString(): String = value
}
