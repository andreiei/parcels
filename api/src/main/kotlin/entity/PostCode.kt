package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class PostCode(private val value: String) {
    override fun toString(): String = value
}
