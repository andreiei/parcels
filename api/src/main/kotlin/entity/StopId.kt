package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class StopId(private val value: String) {
    override fun toString(): String = value
}
