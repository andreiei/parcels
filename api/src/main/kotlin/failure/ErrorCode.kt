package failure

import kotlinx.serialization.Serializable

@Serializable
enum class ErrorCode {
    BAD_REQUEST,
    UNAUTHORIZED,
    NOT_FOUND,
    INTERNAL_SERVER_ERROR,
    FAILED_TO_CREATE_PARCEL,
}
