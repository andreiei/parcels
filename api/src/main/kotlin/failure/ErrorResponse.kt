package failure

import io.netty.handler.codec.http.HttpResponseStatus
import kotlinx.serialization.Serializable

@Serializable
data class ErrorResponse(
    val error: ErrorCode,
    val description: String,
    val code: Int,
) {
    constructor(error: ErrorCode, description: String?, status: HttpResponseStatus) : this(
        error = error,
        description = description ?: status.reasonPhrase(),
        code = status.code(),
    )
}
