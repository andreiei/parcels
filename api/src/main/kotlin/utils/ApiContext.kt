package utils

import services.ParcelRepository
import services.RoutesService

data class ApiContext(
    val parcelRepository: ParcelRepository = ParcelRepository(),
    val routesService: RoutesService = RoutesService(),
)
