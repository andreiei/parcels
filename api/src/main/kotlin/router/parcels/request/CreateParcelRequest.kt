package router.parcels.request

import entity.ParcelId
import entity.PostCode
import kotlinx.serialization.Serializable

@Serializable
data class CreateParcelRequest(
    val id: ParcelId,
    val weight: Double,
    val postCode: PostCode,
    val address: String,
    val status: ParcelStatus,
) {
    fun validate() {
        require(weight in 0.1..1000.0) {
            "Weight must be between 0.1 and 1000.0 kg"
        }

        require(postCode.toString().matches(Regex("\\d{4}"))) {
            "Post code must be a 4-digit number"
        }

        require(address.length in 1..255) {
            "Address must be between 1 and 255 characters"
        }
    }
}

enum class ParcelStatus {
    NOT_STARTED,
    IN_TRANSIT,
    DELIVERED,
}
