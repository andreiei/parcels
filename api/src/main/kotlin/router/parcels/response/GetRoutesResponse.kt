package router.parcels.response

import entity.ParcelId
import entity.RouteId
import entity.StopId
import kotlinx.serialization.Serializable

@Serializable
data class GetRoutesResponse(
    val routes: List<Route>,
)

@Serializable
data class Route(
    val id: RouteId,
    val stops: List<Stop>,
)

@Serializable
data class Stop(
    val id: StopId,
    val address: String,
    val parcels: List<ParcelId>,
)
