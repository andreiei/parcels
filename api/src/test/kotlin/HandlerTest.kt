import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent
import entity.ParcelId
import entity.PostCode
import entity.StopId
import failure.ErrorCode
import failure.ErrorResponse
import io.kotest.assertions.throwables.shouldNotThrow
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.netty.handler.codec.http.HttpMethod
import router.parcels.request.CreateParcelRequest
import router.parcels.request.ParcelStatus
import router.parcels.response.GetRoutesResponse
import utils.ApiContext

class HandlerTest : ShouldSpec({
    val handler = Handler()

    should("call endpoint to create parcel") {
        shouldNotThrow<Exception> {
            handler.handleRequest(
                context = null,
                event = APIGatewayProxyRequestEvent()
                    .withBody(
                        encodeToString(
                            CreateParcelRequest(
                                id = ParcelId("P001"),
                                weight = 2.3,
                                postCode = PostCode("0465"),
                                address = "123 Main St",
                                status = ParcelStatus.IN_TRANSIT,
                            ),
                        ),
                    )
                    .withHttpMethod(HttpMethod.POST.name())
                    .withPath("/v1/parcels"),
            )
        }
    }

    should("call endpoint to fetch today's parcels") {
        val response: APIGatewayProxyResponseEvent = handler.handleRequest(
            context = null,
            event = APIGatewayProxyRequestEvent()
                .withHttpMethod(HttpMethod.GET.name())
                .withPath("/v1/parcels/today"),
        )

        decodeFromString<GetParcelsResponse>(response.body) shouldMatchSnapshot this
    }

    should("call endpoint to fetch today's routes") {
        val response: APIGatewayProxyResponseEvent = handler.handleRequest(
            context = null,
            event = APIGatewayProxyRequestEvent()
                .withHttpMethod(HttpMethod.GET.name())
                .withPath("/v1/parcels/today/routes"),
        )

        decodeFromString<GetRoutesResponse>(response.body)
            .routes
            .map { route ->
                route.copy(stops = route.stops.map { it.copy(id = StopId("REDACTED")) })
            }
            .shouldMatchSnapshot(this)
    }

    should("return a failed response") {
        val localHandler = Handler(
            apiContext = ApiContext(
                parcelRepository = mockk {
                    every { addParcel(any()) } returns Result.failure(RuntimeException("Exception adding parcel"))
                },
            ),
        )
        val response: APIGatewayProxyResponseEvent = localHandler.handleRequest(
            context = null,
            event = APIGatewayProxyRequestEvent()
                .withBody(
                    encodeToString(
                        CreateParcelRequest(
                            id = ParcelId("P001"),
                            weight = 2.3,
                            postCode = PostCode("0465"),
                            address = "123 Main St",
                            status = ParcelStatus.IN_TRANSIT,
                        ),
                    ),
                )
                .withHttpMethod(HttpMethod.POST.name())
                .withPath("/v1/parcels"),
        )

        decodeFromString<ErrorResponse>(response.body).error shouldBe ErrorCode.FAILED_TO_CREATE_PARCEL
    }
})
