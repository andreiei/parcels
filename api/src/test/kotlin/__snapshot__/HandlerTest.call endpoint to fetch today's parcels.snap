{
    "parcels": [
        {
            "id": "P001",
            "status": "IN_TRANSIT",
            "weight": 2.3,
            "postCode": "0465",
            "address": "123 Main St"
        }
    ]
}